src/porepy/numerics/ad/_ad_parser.py

```diff
@@ -0,0 +1,529 @@
+            eq_sys: The EquationSystem wherein the system state is defined.
+
+        Returns:
+            A numpy array or an AdArray representation of the operator op, depending on
+                whether the derivative is requested.
+
+        """
+        # If the operator is in the cache, return the cached value.
+        if op in self._cache:
+            cached = self._cache[op]
+            return cached
+
+        # The operator is not in the cache. Parse the operator by recursion:
+        # 1. If the operator is a leaf (has no children), parse the leaf.
+        # 2. If the operator is a composite operator, parse the children and combine
+        #    them according to the operator.
+        if op.is_leaf():
+            if isinstance(op, pp.ad.MixedDimensionalVariable):
+                if op.is_previous_iterate or op.is_previous_time:
+                    # Empty vector like the global vector of unknowns for prev time/iter
+                    # insert the values at the right dofs and slice.
+                    vals = np.empty_like(
+                        ad_base.val if isinstance(ad_base, pp.ad.AdArray) else ad_base
+                    )
+                    # List of indices for sub variables.
+                    dofs = []
+                    for sub_var in op.sub_vars:
+                        sub_dofs = eq_sys.dofs_of([sub_var])
+                        vals[sub_dofs] = sub_var.parse(eq_sys.mdg)
+                        dofs.append(sub_dofs)
+
+                    return vals[np.hstack(dofs, dtype=int)] if dofs else np.array([])
+                else:
+                    # Fetch the values from the state vector.
+                    return ad_base[eq_sys.dofs_of([op])]
+
+            # Atomic variables.
+            elif isinstance(op, pp.ad.Variable):
+                # If a variable represents a previous iteration or time, parse values.
+                if op.is_previous_iterate or op.is_previous_time:
+                    return op.parse(eq_sys.mdg)
+                # Otherwise use the current time and iteration values.
+                else:
+                    return ad_base[eq_sys.dofs_of([op])]
+            # All other leafs like discretizations or some wrapped data.
+            else:
+                # Mypy complains because the return type of parse is Any.
+                res = op.parse(eq_sys.mdg)  # type:ignore
+                # Profiling indicated that in this case, caching actually pays off, so
+                # we keep it.
+                self._cache[op] = res
+                return res
+
+        # This is not a leaf, but a composite operator. Parse the children and combine
+        # them according to the operator.
+        child_values = [
+            self._evaluate_single(child, ad_base, eq_sys) for child in op.children
+        ]
+
+        # Get the operation represented by op.
+        operation = op.operation
+        match operation:
+            case Operations.add | Operations.sub:
```

Comment:
I somehow get the classification in +/- and */div routines. Though, one could also consider symbol-based and special cases. add seems equally treated as mul. So why was it not possible to integrated sub as special case below as rmatmul?

Reply:
I originally had * together with + and -, but it did not work out with the reversion of child_values. This does no harm for addition, is necessary for subtraction, and is also for necessary for multiplication - but only if the second operand is an AdArray, for other cases it actually does harm. I forgot precisely what was the problem, but it surely was something. 

Hence, the structure ended up as it is. There are a few other structures that would have been functionally equivalent and roughly as logical as this one, but I could not find one superior structure - for that, there are too many special cases to handle.